### PR TITLE
Fix #101: leafletProxy() function calls can get applied out-of-order

### LINKED ISSUE #101: leafletProxy() function calls can get applied out-of-order
```r
leafletProxy("map") %>% clearShapes() %>% addCircles()
```

I'm seeing clearShapes called after addCircles. This is because these function calls get individually deferred using `session$onFlushed` and it turns out those are not guaranteed to be called in the order they were registered.

Ideally this would be fixed in Shiny but we just released to CRAN this week. Another way to go would be to change leafletProxy from a list to a reference class or R6 class, that would at least let us ensure ordering WITHIN a single leafletProxy object (still sucky though).

### PR DESCRIPTION
**Pull request by jcheng5**
cc @wch

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -155,12 +172,45 @@ invokeRemote = function(map, method, args = list()) {
     )
   )
 
+  sess <- map$session
   if (map$deferUntilFlush) {
-    map$session$onFlushed(function() {
-      map$session$sendCustomMessage("leaflet-calls", msg)
-    }, once = TRUE)
+    if (compareVersion(as.character(packageVersion("shiny")), "0.12.0.9001") < 0) {
```
Comment (wch) [inline R/utils.R]: You may know this already, but you can just do `packageVersion("shiny") < "0.12.0.9001"`.

File: R/utils.R [hunk 2]
```diff
@@ -155,12 +172,45 @@ invokeRemote = function(map, method, args = list()) {
...
+
+      if (!exists(sess$token, sessionFlushQueue)) {
+        # If the current session doesn't have an entry in the sessionFlushQueue,
+        # initialize it with a blank list.
+        assign(sess$token, list(), sessionFlushQueue)
+
+        # If the session ends before the next onFlushed call, remove the entry
+        # for this session from the sessionFlushQueue.
+        endedUnreg <- sess$onSessionEnded(function() {
+          rm(list = sess$token, pos = sessionFlushQueue)
```
Comment (wch) [inline R/utils.R]: Better to use `envir` instead of `pos` here.

File: R/utils.R [hunk 3]
```diff
@@ -155,12 +172,45 @@ invokeRemote = function(map, method, args = list()) {
...
-      map$session$sendCustomMessage("leaflet-calls", msg)
-    }, once = TRUE)
+    if (compareVersion(as.character(packageVersion("shiny")), "0.12.0.9001") < 0) {
+
+      # See comment on sessionFlushQueue.
+
+      if (!exists(sess$token, sessionFlushQueue)) {
+        # If the current session doesn't have an entry in the sessionFlushQueue,
+        # initialize it with a blank list.
+        assign(sess$token, list(), sessionFlushQueue)
```
Comment (wch) [inline R/utils.R]: I just want to make sure you're aware that it's much more efficient to do `sessionFlushQueue[[sess$token]] <- list()`.

In general, functions like get/assign/exists are much slower than `[[`. See this thread if you want to know more: 
  http://r.789695.n4.nabble.com/we-need-an-exists-get-hybrid-td4700333.html

**Comment by wch** [comment]
Looks good to me.

**Comment by jcheng5** [comment]
Thanks for the review. Once the Travis tests pass, I'll merge.

**Comment by jcheng5** [comment]
The test I added is failing on Travis, but works for me locally. UGH

**Pull request merged**